From 45e163d4536249aae7c861869473bed52d31bf01 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Mon, 27 Aug 2012 17:37:34 +0300
Subject: [PATCH 300/311] conf: export c2_conf_reg_tl and the accompanying
 functions

Initialisation of c2_conf_obj::co_reg_link will be performed in
conf/obj_ops.c. (It used to be performed in c2_conf_reg_add().)

Previous variant of c2_conf_reg_add() was defective: its precondition
would fail, because c2_tlist_contains() expects initialised tlink.

LogD #840
RB: r/981/diff/1-2/?file=27292#file27292line72
---
 conf/reg.c | 27 ++++++++++++---------------
 conf/reg.h |  7 ++++++-
 2 files changed, 18 insertions(+), 16 deletions(-)

diff --git a/conf/reg.c b/conf/reg.c
index 49cc364..51e297c 100644
--- a/conf/reg.c
+++ b/conf/reg.c
@@ -41,22 +41,20 @@
 
 enum { REG_MAGIC = 0x33fab1edfe0da177 };
 
-C2_TL_DESCR_DEFINE(reg, "registered c2_conf_obj-s", static, struct c2_conf_obj,
-		   co_reg_link, co_magic_gen, C2_CONF_OBJ_MAGIC, REG_MAGIC);
-C2_TL_DEFINE(reg, static, struct c2_conf_obj);
+C2_TL_DESCR_DEFINE(c2_conf_reg, "registered c2_conf_obj-s", ,
+		   struct c2_conf_obj, co_reg_link, co_magic_gen,
+		   C2_CONF_OBJ_MAGIC, REG_MAGIC);
+C2_TL_DEFINE(c2_conf_reg, , struct c2_conf_obj);
 
 void c2_conf_reg_init(struct c2_conf_reg *reg)
 {
-	reg_tlist_init(&reg->r_objs);
+	c2_conf_reg_tlist_init(&reg->r_objs);
 }
 
 void c2_conf_reg_add(struct c2_conf_reg *reg, struct c2_conf_obj *obj)
 {
-	struct objptr *p;
-	C2_PRE(!reg_tlist_contains(&reg->r_objs, obj));
-
-	reg_tlink_init(obj);
-	reg_tlist_add(&reg->r_objs, obj);
+	C2_PRE(!c2_conf_reg_tlist_contains(&reg->r_objs, obj));
+	c2_conf_reg_tlist_add(&reg->r_objs, obj);
 }
 
 struct c2_conf_obj *c2_conf_reg_lookup(const struct c2_conf_reg *reg,
@@ -65,7 +63,7 @@ struct c2_conf_obj *c2_conf_reg_lookup(const struct c2_conf_reg *reg,
 {
 	struct c2_conf_obj *obj;
 
-	c2_tlist_for(&reg_tl, &reg->r_objs, obj) {
+	c2_tlist_for(&c2_conf_reg_tl, &reg->r_objs, obj) {
 		if (obj->co_type == type && c2_conf_buf_eq(&obj->co_id, id))
 			return obj;
 	} c2_tlist_endfor;
@@ -77,15 +75,14 @@ static void del(struct c2_conf_obj *obj)
 {
 	C2_PRE(obj->co_nrefs == 0 && obj->co_status != C2_CS_LOADING);
 
+	c2_conf_reg_tlist_del(obj);
 	if (obj->co_ops->coo_fini != NULL)
 		obj->co_ops->coo_fini(obj);
-
-	reg_tlink_del_fini(obj);
 }
 
 void c2_conf_reg_delete(const struct c2_conf_reg *reg, struct c2_conf_obj *obj)
 {
-	C2_PRE(reg_tlist_contains(&reg->r_objs, obj));
+	C2_PRE(c2_conf_reg_tlist_contains(&reg->r_objs, obj));
 	del(obj);
 }
 
@@ -93,11 +90,11 @@ void c2_conf_reg_fini(struct c2_conf_reg *reg)
 {
 	struct c2_conf_obj *obj;
 
-	c2_tlist_for(&reg_tl, &reg->r_objs, obj) {
+	c2_tlist_for(&c2_conf_reg_tl, &reg->r_objs, obj) {
 		del(obj);
 	} c2_tlist_endfor;
 
-	reg_tlist_fini(&reg->r_objs);
+	c2_conf_reg_tlist_fini(&reg->r_objs);
 }
 
 /** @} conf_dlspec_reg */
diff --git a/conf/reg.h b/conf/reg.h
index ddd9464..a7ec4ba 100644
--- a/conf/reg.h
+++ b/conf/reg.h
@@ -17,11 +17,13 @@
  * Original author: Valery V. Vorotyntsev <valery_vorotyntsev@xyratex.com>
  * Original creation date: 04-Mar-2012
  */
+#pragma once
+
 #ifndef __COLIBRI_CONF_REG_H__
 #define __COLIBRI_CONF_REG_H__
 
 #include "conf/obj.h"  /* c2_conf_objtype */
-#include "lib/tlist.h"
+#include "lib/tlist.h" /* c2_tl, C2_TL_DESCR_DECLARE */
 
 /**
  * @page conf-fspec-reg Registry of Cached Configuration Objects
@@ -73,6 +75,9 @@ struct c2_conf_reg {
 	uint64_t     r_magic;
 };
 
+C2_TL_DESCR_DECLARE(c2_conf_reg, extern);
+C2_TL_DECLARE(c2_conf_reg, , struct c2_conf_obj);
+
 /** Initialises a registry. */
 void c2_conf_reg_init(struct c2_conf_reg *reg);
 
-- 
1.8.3.2

